### DISCUSSION TITLE
What to do with Billing Email?

### DISCUSSION THREAD

**Discussion by tessus** [dani-garcia/vaultwarden#2849] [General]
I was looking mainly for the input of @dani-garcia and @BlackDex, but I think the community might have a few ideas as well.

Currently the use of the Billing Email is a bit inconsistent and in fact not necessary.

- when creating an organization the UI requires for the Billing Email an email address (verified by the input form) (screen 1)
- in `Organizations > Settings > My Organization` the UI allows even an empty string to be set as the Billing Email (screen 2)

Vaultwarden does not require a billing email at all, so here are few ideas:

- hide the billing email field from both screens (and use an empty string)
- hide the billing email field from screen 1 and use the current userid as value
- allow an empty string in screen 1
- rename `Billing Email` to `Primary Owner Email` in both screens (`Primary` because an org _could_ have more than one owner)

The solution could be one or a combination of the above items. I just thought I'd start a discussion, because the current situation is inconsistent and I personally dislike inconsistencies. ;-)

I am also more than happy to make the code changes. That is if we can reach a consensus and I get the go-ahead...

**Comment by bokkabonga** [comment]
As far as i can remember it was possible to not enter anything into the billing email field until the last update. At least no one on my server noticed it until i updated it yesterday.

So i guess that something was changed in that regard (most likely upstream) ?

**Reply by tessus** [reply]
> So i guess that something was changed in that regard (most likely upstream) ?

Yes, which is what I mentioned pretty much in the beginning:

- when creating an organization the UI requires for the Billing Email an email address (verified by the input form) (screen 1)

The discussion I want to have is what we should do about it. ;-) e.g. do you or your users even need this field? If yes, what for?

**Comment by BlackDex** [comment]
As far as i know it has been mandatory since the beginning.
I even checked a very old bitwardenrs image, where I am forced to fill in a password.

I know i have been criticized a few times by mentioning that I personally want to keep as close to Bitwarden it self regarding most stuff, and only deviate where convenient. But I think the same here. This is a mandatory field, and we use it also in the admin interface.

That you currently are able to update it with an empty value is actually an bug. We need to check server-side if that value is empty, and if so, we need to respond with an error, same as Bitwarden does.

The thing is, it might be useful for future items maybe. It might be possible to create some cronjob which sends out mails every day mentioning the amount of ciphers, users, attachments whatever.

Besides that, it isn't that hard to just fill in an email address. Also it means more modifications to the web-vault which I also try to keep to a minimum, mostly because it sometimes takes a lot of effort to patch it.

In my opinion we should just leave it as-is, and even fix the possibility of providing an empty email address at the update form.

**Comment by tessus** [comment]
> I know i have been criticized a few times by mentioning that I personally want to keep as close to Bitwarden it self regarding most stuff, and only deviate where convenient. But I think the same here. This is a mandatory field, and we use it also in the admin interface.

I certainly understand the notion to keep vw as close to Bitwarden as possible. But in this case I think it's different.

When I first created an org, I was rather taken aback, when I read `Billing Email`. It just does not make sense, since there are no invoices, nor is anything sent to that email (it's not verified either).
Removing that part from the admin page is also not very hard (if necessary).

On the other side, I am not saying that we have to get rid of the value, but rather hide it and/or autofill it.

I gave a bunch of options of what _could_ be done.

IMO, the most logical for vaultwarden would be the following:

- change the email input field in screen 1 (creating an org) to a hidden input field and use the currently logged in user (email) as the value (there's no need to fill in something manually and one would most likely not use a different email address if one is using this field in a serious manner)
- change the label in screen 2 (`Organizations > Settings > My Organization`) from `Billing Email` to `Owner Email` or `Primary Owner Email`

This will streamline the workflow and helps also with usability in terms of taking the dreaded billing feel out of the equation.

**Comment by BlackDex** [comment]
Translation is going to be hard, since that needs to be done in all languages then, not only English, so i think that is out of the question.

Hidding can be done, but that takes a bit more invasive patching then only using CSS to hide the field, because JS will still try to validate that field. And i actually want to try and prevent modifications as much as possible as mentioned before.

If only hiding it would have helped, then probably no problem, but it isn't. And dynamically hiding it is going to be a pain also.

Though there is something to say on hiding it, but i actually think it can still be useful, especially in larger setups.

**Reply by tessus** [reply]
> Translation is going to be hard,

There's a translation entry for `Owner` and `Email`, so it should be fine.

> Hidding can be done, but that takes a bit more invasive patching then only using CSS to hide the field, because JS will still try to validate that field. 

This should also not be a problem, if the field is pre-filled with the email address. it's hidden but pre-filled, so the validation should also be ok.

I am only saying to hide it in the "create org" screen, not in the settings screen.

But even if we don't hide it, but only pre-fill it, it should be a better experience.

When I have a bit more time, I'll try to make these changes with as little patching as possible. You can still decide then.

Having said all that, I really understand your concerns and I certainly do not want to create a maintenance burden. I am just saying that if there's a chance that this can be done easily, it might be worth it.

**Reply by BlackDex** [reply]
Pre-fill it how? We can't do that, at least not without any yet again more modifications to the web-vault.

adjusting that specific field to not be mandatory is not that hard i think, but still again more files to patch and update the patching when Bitwarden makes more intensive changes or reorder files which they have done several times.

Also, if hiding that field, you need to change the rust code also to allow that field to be missing.

But, again, i think it still makes sense in some environments to have tha field. And i actually do not see the big issue here. It's not that you need to register 175 organizations one at a time right after each other.

**Reply by tessus** [reply]
> Pre-fill it how?

With something like this. I still haven't had time to test it, but by browsing through the code this came to mind:

```
    <bit-form-field class="tw-w-1/2">
      <bit-label>{{ "owner" | i18n }} {{ "email" | i18n }}</bit-label>
      <input bitInput type="hidden" formControlName="billingEmail" [(ngModel)]="profile.email" />
    </bit-form-field>
```

PS.: The above label has to be removed. I just did that to show how to use available translation.

> Also, if hiding that field, you need to change the rust code also to allow that field to be missing.

No, I think you misunderstand. Hiding an input field does not remove the value when the form is submitted.

```
<form action="#">
    <input type="text" name="org">
    <input type="hidden" name="email" value="test@example.com">
    <input type="submit" value ="submit">
</form>
``` 

> i think it still makes sense in some environments to have tha field.

I am not saying it doesn't. I said the field should be hidden when **creating an organization**, not when changing the settings of an org.

**Reply by BlackDex** [reply]
I understood it correctly, but if you hide it in screen 1, nobody is going to screen 2 to fill it in there, that would be a bad flow also. Only thing what would be an option is not having it mandatory, but then we are probably going to need a new config option, again, to configure if you want it mandatory or not.

Using two different translated words is a bad idea also. Who says the order of the words are correct, or that when combined the spelling is different, language is not that easy ;).

Also, adding a dummy default is also bad I think, that would clutter the database with invalid items, maybe people questioning where it came from etc..

Don't understand me wrong, i see what you mean. But, to be honest, how much work is filling in some email address?

**Reply by BlackDex** [reply]
Maybe only email as translation would be enough? But again, that would be more intensive changes which are prone to much changes if they charge the layout again, which brings again more maintenance burden.

**Reply by tessus** [reply]
> nobody is going to screen 2 to fill it in there

Why would somebody fill it in there? Their email (or the person's who created the org) is already in there. People would only go to that screen, if they wanted to change the value.

> adding a dummy default is also bad I think

I never said to use a dummy default. I said the current userid (email) will be used. When I am logged in as `myuser@mydomain.com` the billing email (which would be renamed to email) would be automatically set to `myuser@mydomain.com` (and hidden).

> how much work is filling in some email address?

The point is that it is confusing having to fill in a billing email when creating an org. It's not that you would break your fingers by typing your (or any) email address.

> Using two different translated words is a bad idea also. Who says the order of the words are correct, or that when combined the spelling is different, language is not that easy ;).

Yes, you are correct, but as a quick hack it could be sufficient. I am not saying it is the final solution. As you pointed out, just using `Email` would be enough.

Maybe I'll create a POC and you can see what I mean.

**Reply by tessus** [reply]
If it is a too complex change, I wouldn't merge it either.

**Comment by tessus** [comment]
I was able to write less than 5 lines of javascript to fill in the email (current userid), but I can't get it to work with AngularJS. I am not a web guy, so I could need a bit of help there.

While I understand that hiding the field (in the create org screen) might not be wanted, it could still be convenient to have the email field pre-populated with your email address.

I believe that this part of the code was already refactored by the Bitwarden team, thus I doubt that major changes will happen there. This means that maintenance effort for such a patch should be low.

**Reply by BlackDex** [reply]
Then still, we need to maintaine it every update. It also doesn't matter if they update that specific part, they could also update code above our near which could cause the patch to fail.

And, i think that this is a change which change the interface automatically, which in my opinion is a nogo to add from our side.

**Reply by tessus** [reply]
Even if this is never going to be merged, I can still change it on my own instance. ;-)

Personally I find it a major UI flaw having to fill in an email address when I create an org. In Vaultwarden that is. In Bitwarden it can make sense not to pre-populate the field, but in vaultwarden this field is utterly useless and if one where to use it, most likely noone would use sombody else's email address.

**Reply by stefan0xC** [reply]
While the naming is not the best (since I don't charge my users anything), I do think that there is a reason for not pre-populating the field with a personal email address. At least in my use-case I want my users to fill in their group email address and not their personal address when creating an organization since then I can see which group an organization belongs to instead and they are free to make up their organization names as it makes sense to them. If it's prefilled the likelihood that it gets filled out correctly will drop significantly (and I had to ask each organization owner to change the email address regularly).

btw: The check that was added in my PR #2864 is not very strict (e.g. you can just add `a@b` if you don't want to type too much.)

**Reply by tessus** [reply]
I use this field to see who created an org. This person is most likely the owner. Should I have to inquire something, I can send that person an email.

**Reply by tessus** [reply]
I have already understood that pre-populating the billing email field won't happen in Vaultwarden.

So let's go back to the basics. Any chance we can remove the word billing? Is a PR going to be accepted for that?

**Reply by stefan0xC** [reply]
The change for one or two languages should be an easy change (e.g. by changing the value of the field `billingEmail` in `apps/web/src/locales/en/messages.json` or `web-vault/locales/en/messages.json`) which you can do in your installation but (as BlackDex already [pointed out](#discussioncomment-3937547)) I am not sure if it is feasible for Vaultwarden as we would have to change (and maintain) the translation of (currently) [55 languages](https://bitwarden.com/help/localization/).

**Reply by tessus** [reply]
No, that's not necessary at all. Thi sis only nrcessary if we were to change a translation or want to use multiple words. email is already translated.

**Reply by tessus** [reply]
I'll open a PR an you can see what I mean.

**Reply by stefan0xC** [reply]
Ah, I okay. So you want the two uses of the translated `"billingEmail" | i18n` to be changed to `"email" | i18n`, I think that should be an easy change if we can live with such a small deviation. 

// Removed because the same suggested change can be discussed in the PR //

**Reply by tessus** [reply]
yep, https://github.com/dani-garcia/bw_web_builds/pull/92
